engine/src/main/java/io/zeebe/engine/nwe/behavior/BpmnEventSubscriptionBehavior.java

```diff
@@ -0,0 +1,199 @@
+ */
+package io.zeebe.engine.nwe.behavior;
+
+import io.zeebe.engine.nwe.BpmnElementContext;
+import io.zeebe.engine.processor.Failure;
+import io.zeebe.engine.processor.KeyGenerator;
+import io.zeebe.engine.processor.TypedStreamWriter;
+import io.zeebe.engine.processor.workflow.CatchEventBehavior;
+import io.zeebe.engine.processor.workflow.ExpressionProcessor.EvaluationException;
+import io.zeebe.engine.processor.workflow.deployment.model.element.ExecutableActivity;
+import io.zeebe.engine.processor.workflow.deployment.model.element.ExecutableBoundaryEvent;
+import io.zeebe.engine.processor.workflow.deployment.model.element.ExecutableCatchEventSupplier;
+import io.zeebe.engine.processor.workflow.message.MessageCorrelationKeyException;
+import io.zeebe.engine.state.ZeebeState;
+import io.zeebe.engine.state.instance.ElementInstance;
+import io.zeebe.engine.state.instance.ElementInstanceState;
+import io.zeebe.engine.state.instance.EventScopeInstanceState;
+import io.zeebe.engine.state.instance.EventTrigger;
+import io.zeebe.engine.state.instance.StoredRecord.Purpose;
+import io.zeebe.protocol.impl.record.value.workflowinstance.WorkflowInstanceRecord;
+import io.zeebe.protocol.record.intent.WorkflowInstanceIntent;
+import io.zeebe.protocol.record.value.BpmnElementType;
+import io.zeebe.protocol.record.value.ErrorType;
+import io.zeebe.util.Either;
+import io.zeebe.util.buffer.BufferUtil;
+
+public final class BpmnEventSubscriptionBehavior {
+
+  private final WorkflowInstanceRecord eventRecord = new WorkflowInstanceRecord();
+
+  private final BpmnStateBehavior stateBehavior;
+  private final BpmnStateTransitionBehavior stateTransitionBehavior;
+  private final EventScopeInstanceState eventScopeInstanceState;
+  private final ElementInstanceState elementInstanceState;
+  private final CatchEventBehavior catchEventBehavior;
+  private final TypedStreamWriter streamWriter;
+  private final KeyGenerator keyGenerator;
+
+  public BpmnEventSubscriptionBehavior(
+      final BpmnStateBehavior stateBehavior,
+      final BpmnStateTransitionBehavior stateTransitionBehavior,
+      final CatchEventBehavior catchEventBehavior,
+      final TypedStreamWriter streamWriter,
+      final ZeebeState zeebeState) {
+    this.stateBehavior = stateBehavior;
+    this.stateTransitionBehavior = stateTransitionBehavior;
+    this.catchEventBehavior = catchEventBehavior;
+    this.streamWriter = streamWriter;
+
+    eventScopeInstanceState = zeebeState.getWorkflowState().getEventScopeInstanceState();
+    elementInstanceState = zeebeState.getWorkflowState().getElementInstanceState();
+    keyGenerator = zeebeState.getKeyGenerator();
+  }
+
+  public <T extends ExecutableActivity> void triggerBoundaryEvent(
+      final T element, final BpmnElementContext context) {
+    final var eventTrigger =
+        eventScopeInstanceState.peekEventTrigger(context.getElementInstanceKey());
+
+    if (eventTrigger == null) {
```

Comment:
can we have comment why this could happen?

Reply:
I think the event trigger is removed when the element (i.e. its event scope) is left.